### DISCUSSION TITLE
Real-time Path Layer Update

### DISCUSSION THREAD

**Discussion by AlexandreL9** [visgl/deck.gl#7068] [Q&A]
I am trying to use the path-layer of DeckGL in real-time, in my case, I have thousands of vehicles that transmit their positions every second, and these are available on a server.
So currently DeckGL path-layer will take props.data and then draw() the trajectory segments on the map (with WebGL...), if props.data changes in size, then all the trajectory is redraw(), but if in the same props.data only a part changes, then we redraw() only those positions are updated

I made a test using the dataset of the cabs of New York. My DeckGL code fetches the JSON with all the last 60 positions from my server and will copy it into data.props, I know it's not optimal, it's just to test.

![taxi_test](https://user-images.githubusercontent.com/32158297/174905926-2aca245c-d2b9-4123-9e6d-db65e6756635.png)

In the end, it works well on my laptop to display smoothly (30 fps) the last 60 positions/seconds of 5000 vehicles.

The problem with the actual implementation is that it means updating the last position of the array (props.data), so we have to shift everything to the left, delete the oldest position and then append the most recent one at the end. So in the end, we have to update the whole array of positions. A logical solution would be a chained list, but the implementation of Deckgl is incompatible with that.
The other "custom" solution would be to create a new layer every second (without exceeding 256 layers/seconds) and delete the oldest one.
But the most optimal would be that the redraw() takes into account not the whole array(props.data) of positions but just the new position and deletes the oldest fragment of the path. Still, I hardly see how to do it, and it would be very complex to implement.

Is there another approach to handle real-time for this scenario ?

**Comment by Pessimistress** [comment]
Incremental append is supported in `PathLayer` in the sense that you can add additional paths to the data array, but not change the length of each path. You can either supply an [asynchronous iterable](https://deck.gl/docs/api-reference/core/layer#_datadiff) to `props.data`, or implement [_dataDiff](https://deck.gl/docs/api-reference/core/layer#_datadiff) yourself if you wish to insert chunks instead of appending them to the end.

**Reply by AlexandreL9** [reply]
I'm trying to use an async iterator as in [this solution](https://github.com/visgl/deck.gl/issues/3838) 
In my case, the JSON that I fetch will be up-to-date with last positions, and I will keep only those greater than the last fetch timestamp
```
import React, {useState, useEffect} from 'react';
import {render} from 'react-dom';
import {StaticMap} from 'react-map-gl';
import DeckGL from '@deck.gl/react';
import {PathLayer} from '@deck.gl/layers';


const INITIAL_VIEW_STATE = {
  longitude: -73.90576,
  latitude: 40.61,
  zoom: 12,
};

const MAP_STYLE = 'https://basemaps.cartocdn.com/gl/voyager-gl-style/style.json';

export default function App({
  initialViewState = INITIAL_VIEW_STATE,
  mapStyle = MAP_STYLE,
}) {
  
	function timeout(ms) {
		return new Promise(resolve => setTimeout(resolve, ms));
	}
	
	function fetch_data(){			
		fetch("https://raw.githubusercontent.com/visgl/deck.gl-data/master/examples/trips/trips-v7.json")  
      .then((resp) => resp.json()) 
      .then((resp) => {
		return resp;
		});
	}

	async function* getData() {
		while(1){
			await timeout(1000);
			yield [fetch_data()];
		}
	}

    function updateLayers(props) {
		const { paths } = props;
		const layers =[
		  new PathLayer({
			  id: 'PathLayer',
			  data: paths,
			  getPath: d => d.path
			})
		];
		return layers;
	}
		
	return (
	<DeckGL
	  layers={updateLayers({data: getData()})}
	  initialViewState={initialViewState}
	  controller={true}
	>
	  <StaticMap reuseMaps mapStyle={mapStyle} preventStyleDiffing={true} />
	</DeckGL>
	);
}

export function renderToDOM(container) {
  render(<div><App /></div>, container);
}
```


However, with this code I obtain this error
```
deck.js:107 TypeError: update of PathLayer({id: 'PathLayer'}): Cannot read properties of undefined (reading 'path')
    at getPath (app.js:57:23)
    at PathTesselator._forEachGeometry (tesselator.js:130:24)
    at PathTesselator._rebuildGeometry (tesselator.js:161:12)
    at PathTesselator.updateGeometry (tesselator.js:58:14)
    at PathLayer.updateState (path-layer.js:160:22)
    at PathLayer._updateState (layer.js:732:14)
    at PathLayer._update (layer.js:712:12)
    at LayerManager._updateLayer (layer-manager.js:272:13)
    at LayerManager._updateSublayersRecursively (layer-manager.js:225:16)
    at LayerManager._updateLayers (layer-manager.js:186:10)
```

Is it a bug, or I do it in the wrong way ?

**Reply by Pessimistress** [reply]
You are passing `undefined`  to `data`.

**Reply by AlexandreL9** [reply]
My asynchronous iterator, finally works, I read a JSON, it adds path segments every second, but I wanted to do a _dataDiff after each retrieval to add only the positions that are not already in the path, to avoid the segments overlapping like now:
 ![image](https://user-images.githubusercontent.com/32158297/176471192-a003fea5-427b-418a-bace-0044d5e73859.png)

However, as I read in the documentation 
"This becomes trickier when dealing with dynamic length data, for example PathLayer, PolygonLayer and GeoJsonLayer. In general, it is not recommended to use this feature when the number of vertices in the paths/polygons may change."

Is it a good idea a solution where I have a list, where for each path, I have the timestamp of the last update, so I only keep the new position from my fetch() response in my getData() function?

What is the most efficient way to remove old positions in my asynchronous iterator after a certain time?

```
/* global window */
import React, {useState, useEffect} from 'react';
import {render} from 'react-dom';
import {StaticMap} from 'react-map-gl';
import DeckGL from '@deck.gl/react';
import {PathLayer} from '@deck.gl/layers';
import {IconLayer} from '@deck.gl/layers';

const INITIAL_VIEW_STATE = {
  longitude: -73.90576,
  latitude: 40.61,
  zoom: 12,
};

const _THEME = [
[100, 250, 93],
[23, 165, 137],
[100, 250, 150],
[100, 150, 255],
[250, 150, 250],
[255, 50, 50]
];

const MAP_STYLE = 'https://basemaps.cartocdn.com/gl/voyager-gl-style/style.json';

export default function App({
  initialViewState = INITIAL_VIEW_STATE,
  mapStyle = MAP_STYLE,
	}) {
	function timeout(ms) {
		return new Promise(resolve => setTimeout(resolve, ms));
	}
	
	async function* getData() {
		var resp = JSON.parse('[ { "vendor": null, "path": [ ], "timestamps": [ ] } ]');
		
		while(1){
			fetch("http://FILE_WITH_LAST_POSITIONS.json")
				.then((response) => response.json()) 
				.then((response) => {
					if (Object.keys(response).length>0){
						resp = response;
					}
			});	
			await timeout(1000);
			yield resp;
		}
	}

    const dataVar = {data: getData()};

    function updateLayers() {
		
		const { data } = dataVar;
		const layers =[
		  new PathLayer({
			  id: 'PathLayer',
			  data,
			  getPath: d => d.path,
			  getWidth: d => 2,
			  widthMinPixels: 2,
			  widthScale: 2,
			  getColor: d => (_THEME[(Math.floor(Date.now() / 1000))%6]),
			})
		];
		return layers;
	}
	return (
	<DeckGL
	  layers={updateLayers()}
	  initialViewState={initialViewState}
	  controller={true}
	>
	  <StaticMap reuseMaps mapStyle={mapStyle} preventStyleDiffing={true} />
	</DeckGL>
	);
}

export function renderToDOM(container) {
  render(<div><App /></div>, container);	
}
```

**Reply by Pessimistress** [reply]
You need to make your async iterator return ONLY the new segments, not the part that is already rendered. The quote from the documentation means that you cannot add vertices to an existing path object, but you can always just add another path that is the new leg.
